### PR TITLE
Implement compound assignment operators for BitSet

### LINKED ISSUE #24: Combine bitsets
Would be good if we had a way to non-lazily combine 2 bitsets in a performant way.

Naive way is just:
```rust
for bit in bitset {
    other.add(bit);
}
```

### PR DESCRIPTION
**Pull request by WaDelma**
Resolves #24

---
This change is [Reviewable](https://reviewable.io/reviews/slide-rs/hibitset/28)

### CODE REVIEW

**Review by torkleyy — COMMENTED** [review]
I like it! I think it's correct, but I want to take a closer look at it. Don't you like my idea (using compound assignment operators)?

**Comment by WaDelma** [comment]
I actually didn't understand what you meant by it and I already had this idea in my head.

**Comment by torkleyy** [comment]
Just using |= for the operation.

**Comment by WaDelma** [comment]
Ah yeah. I can add the trait implementation.

**Comment by WaDelma** [comment]
Hmm.. Should I just implement `&=` and `^=` too then?

**Comment by torkleyy** [comment]
That's what I would expect before closing the issue. If you only want to implement `|=`, that's fine, too.

**Comment by WaDelma** [comment]
I implemented `|=` (which was basically just moving `add_all`) and `&=` which was bit more involved.

To implement `&=` I added `clear` method for `BitIter`. ~~Should we use that for the `BitSet::clear`? As far I can see it should perform better, right? Have to benchmark that.~~

There is still some problems with the PR. I had to duplicate `handle_level` method. I'll try to clean it up.

**Comment by WaDelma** [comment]
Also `&=` has efficiency problem: It always grows the rhs. I'll have to think how to fix that.

**Comment by WaDelma** [comment]
I think this might be ready for a review.

**Review by torkleyy — COMMENTED** [review]
I like the code, it's just that I didn't review the logic yet. I'm not sure when I will be able to do so.

**Comment by WaDelma** [comment]
So is there anything to improve here?
(Mostly just pinging this so it won't be forgotten)

**Review by torkleyy — APPROVED** [review]
Thanks! Sorry, I forgot this one.

**Comment by kvark** [comment]
Would be nice to have better commit messages

File: src/iter/mod.rs [hunk 1]
```diff
@@ -38,7 +38,29 @@ impl<T: BitSetLike> BitIter<T> {
     }
 }
 
-enum State {
+impl<'a> BitIter<&'a mut BitSet> {
+    /// Clears the rest of the bitset that hasn't been iterated yet.
+    pub fn clear(&mut self) {
+        use self::State::*;
+        'find: loop {
```
Comment (kvark) [inline src/iter/mod.rs]: could this be rewritten as
```rust
while let Some(level) = (1..LAYERS)..find(|&level| self.handle_level(level) == Continue) {
    let lower = level - 1;
    let idx = (self.prefix[lower] >> BITS) as usize;
    *self.set.layer_mut(lower, idx) = 0;
}
```

**Comment by torkleyy** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/slide-rs/hibitset/28)*

File: src/ops.rs [hunk 2]
```diff
@@ -10,23 +10,12 @@ impl<'a, B> BitOrAssign<&'a B> for BitSet
...
-                        let idx = iter.prefix[lower] as usize >> BITS;
-
-                        *self.layer_mut(lower, idx) |= lhs.get_from_layer(lower, idx);
-                        continue 'find;
-                    }
-                    Empty => {},
-                }
-            }
-            break;
+        while let Some(level) = (1..LAYERS).find(|&level| iter.handle_level(level) == Continue) {
```
Comment (kvark) [inline src/ops.rs]: wow, glad to see this working. Looks better!

File: src/iter/mod.rs [hunk 3]
```diff
@@ -111,3 +105,30 @@ impl<T: BitSetLike> BitIter<T> {
...
+            assert_eq!(0, i);
+        }
+        for &i in &set.layer1 {
+            assert_eq!(0, i);
+        }
+        for &i in &set.layer0 {
+            assert_eq!(0, i);
+        }
+    }
+}
```
Comment (kvark) [inline src/iter/mod.rs]: nit: newline missing

**Review by kvark — APPROVED** [review]
last commit is better squashed, otherwise good to go

**Comment by torkleyy** [comment]
bors delegate+

**Comment by WaDelma** [comment]
bors r=torkleyy

**Pull request merged**